allennlp/data/semparse/type_declarations/type_declaration.py

```diff
@@ -396,18 +420,33 @@ def get_valid_actions(name_mapping: Dict[str, str],
         # Type to terminal productions.
         for substituted_type in _substitute_any_type(name_type, basic_types):
             valid_actions[str(substituted_type)].add(_make_production_string(substituted_type, name))
+            print(str(substituted_type))
         # Keeping track of complex types.
         if isinstance(name_type, ComplexType) and name_type != ANY_TYPE:
             complex_types.add(name_type)
 
+    print('COMPLEX TYPES')
     for complex_type in complex_types:
         if isinstance(complex_type, PlaceholderType):
+            print(complex_type)
             _get_placeholder_actions(complex_type, basic_types, valid_actions)
         else:
+            # We can produce complex types with a lambda expression, though we'll leave out
+            # placeholder types for now.
+            for i in range(num_nested_lambdas):
+                lambda_var = chr(ord('x') + i)
+                production_string = _make_production_string(complex_type.first,
+                                                            ['lambda ' + lambda_var, complex_type.second])
+                valid_actions[str(complex_type)].add(production_string)
```

Comment:
If you run the tests from this branch (or look at the printed output on travis, probably), you can see the action sequences we're trying to produce.  We want to give action ids for the actions in the target sequence, where the ids are positions in the list of all actions.  But there's an action that is `<r,d> -> ['lambda x', d]`.  The trouble is that `<r,d>` isn't in the type declaration; it comes from `<d,r> -> [<<#1,#2>,<#2,#1>>, <r,d>]`.  So how do I get that rule into the list of actions?

A couple of options:

1. Keep track of the actual type object when creating the `valid_actions` dict, and just add a `lambda` to every key that's a complex type (or maybe every key?).  We need to keep track of the type object because recovering the input and output types from the type string is a pain.
2. Have a mapping from production rule string to something like `LAMBDA` and `VAR` for the lambda and var rules, and have those not depend on the type of the left hand side.

Any other options?  I looked briefly through the data, and it looked like the only place that lambdas showed up was right after `reverse` - it could be that we could hard code something here that would simplify things a bit.

Reply:
I'm going to pursue option 1 for now.

Reply:
I think it would be easy to add `lambda` rules to `type_declaration.get_valid_actions` directly. And I agree that `lambda`s occur in a small set of contexts, but it could be simpler than what you described. I've seen them only within `argmax` and `argmin` where they are used to extract the attributes that need to be maximized or minimized. This means that the productions are always of the form `<b,d> -> ['lambda x', d]`, where `b` is one of the basic types (I don't think it is ever a complex type), and the return type in the head of the production rule is always `d`.

Reply:
I've seen them in a few other places, too.  You can see in the current code in this PR how I implemented this, if you want to look at it now.  I'm pretty close to done - just need to finish the action embedding, and put in a few pieces to keep track of lambdas.  Probably one more day of work before I can run the model on the full day.  Hopefully...